### PR TITLE
Fix behaviour of deerload with 2D-datasets 

### LINKED ISSUE #82: Deerload 2D dataset dimension mapping
Hey,
I encounter a problem using deerload. Using this script

```python
dataImport = dict() # introduce as dictionary
dataImport['xyAxisArray'],dataImport['intensity'],dataImport['parameters'] = dl.deerload(filepath,full_output=True)
```
I get returned the correct data. When taking the xyAxisArray apart I receive the correct axises.
The dimension of the intensity array matches, but the values seem to be populated the wrong direction. So my data get shiftet with every scan (here time vs scan vise).
Data is included for testing, working on a fix right now.

[20200621_DB261_theta-spatial.zip](https://github.com/JeschkeLab/DeerLab/files/6061987/20200621_DB261_theta-spatial.zip)

### PR DESCRIPTION
**Pull request by laenan8466**
The resolves the error that the data was mapped the wrong direction.
Error was found for the attached dataset and need testing for further 2d datasets.
[2D_ScanTime.zip](https://github.com/JeschkeLab/DeerLab/files/6062181/2D_ScanTime.zip)

Fixes #82

### CODE REVIEW

**Comment by luisfabib** [comment]
I have locally run some diagnostics and the bug seems to be somewhere else. However, before I push anything since I do not know exactly what to expect from the data: @laenan8466 are these plots what you would expect from the dataset (with dimensions 285x300)?

issue82

**Comment by laenan8466** [comment]
Hey, nope wrong direction.
The time axis has 285 entries and there are 300 scans (from which only 111 contain data).
This script plots them for me (with my fix):
```python
# Load data
dataImport = dict() # introduce as dictionary
dataImport['xyAxisArray'],dataImport['intensity'],dataImport['parameters'] = dl.deerload(filepath,full_output=True)

# Convert into data array
data2d = dict() # dictonary
data2d['time']=dataImport['xyAxisArray'][:,0]*1e-6 # Get first column with time and convert to seconds
# remove nan values using numpy.isnan() 
# and numpy.logical_not 
data2d['time'] = data2d['time'][np.logical_not(np.isnan(data2d['time']))] 

data2d['scans'] = np.round(dataImport['xyAxisArray'][:,1]*1000+1) # convert scan id to 1 ... n and as integer

# Delete all zero elements
# mask them
zeroMask = np.ma.masked_equal(dataImport['intensity'],0) # mask all values that are zero

data2d['intensity']=dataImport['intensity'][~zeroMask.mask[:,0],:] # use only row mask to delete all rows (=scans) that are zero
data2d['scans']=data2d['scans'][~zeroMask.mask[:,0]] # also crop scans


plt.plot(data2d['time'],np.transpose(data2d['intensity']))
```

grafik

For reference: You are looking at a series of LaserIMD data. So there is a plateau in the beginning, a decay and increase and then a plateau again.

**Comment by luisfabib** [comment]
Ok, now I understand. Thanks, I'll keep working on it.

**Comment by laenan8466** [comment]
As reference - if you sum up all data in scan direction one should result s.th. like this:
grafik

**Comment by luisfabib** [comment]
The cause of this bug was a line inside the file reader code block that was creating a too large matrix leading to a bug-ish reshaping later. 

Additionally, I have improved the interface for the `abscissa` output. Before N abscissa with M being the largest amount of points was being returned as NxM matrix. The shorter abscissas were filled with NaN values. This is a MATLAB relic from the migration and makes no sense in the context of Python. Now, `abscissa` is a list of arrays. Each n-th element of the list is the abscissa array of the corresponding n-th dimension of the dataset. 

Before: 
````python
t1 = abscissa[:,0]
t2 = abscissa[:,1]
t2 = t2[~np.isnan(t2)]
``````
Now: 
````python
t1 = abscissa[0]
t2 = abscissa[1]
``````

**Comment by luisfabib** [comment]
Oh, it seems I had a bug in my test. The reshaping error still persists.

**Comment by laenan8466** [comment]
I second the new abscissa format. Good choice!

Just text me, if I can assist with the reshaping/import errror.

**Comment by luisfabib** [comment]
The issue seems to be on the definition of the dimensions. The data are read as a 1D-array and split according to the XPTS and YPTS defined in the DSC file to form an XPTSxYPTS matrix. In this case, the dataset should be 300x285, however, in the DSC file XPTS=285 and YPTS=300.

**Comment by laenan8466** [comment]
I see.
Does the loading fits with other 2D-datasets that doesn't use 'scans' on one axis (YNAM= 'unknown')? Do you have another dataset with multiple scans to compare?

Not sure why this problem occurs but I remember similar issues when working with a MATLAB importer library we wrote for our group. Our solution was to analyse the X/YNAM and assign different importers. But parts of our import depended on @stestoll easyspin eprload.

**Comment by luisfabib** [comment]
I think I have fixed the problem with the last set of commits. With this MWE:
````python
import numpy as np 
import matplotlib.pyplot as plt 
import deerlab as dl 


dataImport = dict() # introduce as dictionary
x,y,par = dl.deerload('LaserIMD_01_DB261_20200621.DSC',full_output=True)

plt.subplot(211)
plt.plot(x[0],np.sum(y,axis=1))
plt.ylabel('V(t1)')
plt.xlabel('t1 [us]')
plt.tight_layout()
plt.show()
```` 
I obtain the correct accumulated signal as @laenan8466 described above:
issue82 

I can obtain satisfactory results as well with other 2D-datasets.

**Review by luisfabib — APPROVED** [review]

File: deerlab/deerload.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def deerload(fullbasename, plot=False, full_output=False, *args,**kwargs):
     parameters = read_description_file(filename_dsc)
     parDESC = parameters["DESC"]
     parSPL = parameters["SPL"]
-    
+    print( int(parDESC['XPTS']))
```
Comment (laenan8466) [inline deerlab/deerload.py:66]: should this be removed bc. it is debug only?
```suggestion
    

```

**Review by laenan8466 — COMMENTED** [review]
Works for me, suggestions in comments above (minor).
Thanks for all the work!

**Comment by luisfabib** [comment]
Seems I forgot a couple of `print` there.... Nice catch! Thanks!

**Review by stestoll — APPROVED** [review]

**Review by luisfabib — APPROVED** [review]

**Pull request merged**